### ISSUE TITLE
Genomic ranges indexing costly

### ISSUE THREAD

**Issue by LiNk-NY** [vjcitn/MultiAssayExperiment#103]
In order to create names for each genomic range, we thought that adding a numeric character index to each range (within each element of a GRangesList/RangedRaggedAssay) would be good. Unfortunately, this operation turns out to be costly the way it is implemented. 

Please see: 
https://github.com/LiNk-NY/MultiAssayExperiment/blob/master/R/MultiAssayExperiment.R#L4

Any suggestions on how to remedy this? 

Regards,
MR

**Comment by vjcitn** [comment]
I don't quite undertstand where this is going.  after
example(RangedSummarizedExperiment)

> rse

class: RangedSummarizedExperiment

dim: 200 6

metadata(0):

assays(1): counts

rownames: NULL

rowData names(1): feature_id

colnames(6): A B ... E F

colData names(1): Treatment

> names(rse)

NULL

> validObject(rse)

[1] TRUE


> names(rse) = "A"

> validObject(rse)

[1] TRUE

> names(rse) = c("A", "B")

> validObject(rse)

[1] TRUE


so it seems to me the API for names() for RangedSummarizedExperiment is not
specified; it

is not given in the man page.  What is the intent?  names(rowRanges()) i
can understand, and if you

want to require it to meet certain conditions you can force this, but
names() seems too vague.


if managing names(rowRanges()) leads to inefficiencies that would seem to
be a problem for

the SummarizedExperiment maintainers.

**Comment by LiNk-NY** [comment]
Hi Vince, 
We are subsetting using character vectors. In order to find compatible ranges, we take the names of those ranges that are valid and use those to subset in subsequent operations. An example list that is used to subset the Elist in the MultiAssayExperiment can be loaded as: 

```
example("getHits")
```
In the event that either RangedSummarizedExperiment or RangedRaggedAssay objects do not have names (ie. rownames(rse), rownames(RRA)), we impute simple names in order to carry out these subsetting operations. 
This imputation method proves to be costly the way it is implemented.

**Comment by vjcitn** [comment]
On Tue, Feb 16, 2016 at 8:55 PM, Marcel Ramos <notifications@github.com>
wrote:

> Hi Vince,
> We are subsetting using character vectors. In order to find compatible
> ranges, we take the names of those ranges that are valid and use those to
> subset in subsequent operations. An example list that is used to subset the
> Elist in the MultiAssayExperiment can be loaded as:
>
If I understand correctly this entails that the RangedSummarizedExperiment
API (and others) must be elaborated so that the behavior of names() is
specified.  As I noted previously this does not seem to be the case.  The
MAE API should also be updated to reflect this requirement -- that names(x)
provide some effective labeling of features.

I am responding to
} else if (is(object, "RangedSummarizedExperiment")) {
names(object) <- 1:length(object)

> example("getHits")
>
> In the event that either RangedSummarizedExperiment or RangedRaggedAssay
> objects do not have names (ie. rownames(rse), rownames(RRA)), we impute
> simple names in order to carry out these subsetting operations.
> This imputation method proves to be costly the way it is implemented.
>

Maybe I am being pedantic but rownames() is part of the MAE API while
names() is not.  Requiring a non-null rownames seems reasonable to me, and
if something comes in without it I would suggest throwing an error.
Concocting names doesn't seem to be such a good idea.

Also it looks like the names on a GRangesList (if present) would be
clobbered in .createNames?

**Comment by mtmorgan** [comment]
names() would be really confusing here -- SE implements matrix-like API, so rownames()/colnames(). names() on a data.frame() would be columns, but the suggestion here is that it would be 'features' == rows ??


________________________________

**Comment by lwaldron** [comment]
To summarize what I think Vince and Martin are saying here, and I agree: 

1. rownames() is part of the API and names() is not, so take names() out of the code. 
2. add validity checks that rownames() and colnames() must be non-NULL. Don't Generated rownames or colnames are unlikely to be useful.

On a related note, any thoughts about the Issue #98 suggestion for subsetting by rows for the RangedRaggedAssay?

**Comment by PeteHaitch** [comment]
@mtmorgan: SummarizedExperiment currently has `names()` acting like `rownames()`:

```r
> selectMethod('names', 'SummarizedExperiment')
Method Definition:

function (x) 
x@NAMES
<environment: namespace:SummarizedExperiment>

Signatures:
        x                     
target  "SummarizedExperiment"
defined "SummarizedExperiment"
> selectMethod('names', 'RangedSummarizedExperiment')
Method Definition:

function (x) 
names(rowRanges(x))
<environment: namespace:SummarizedExperiment>

Signatures:
        x                           
target  "RangedSummarizedExperiment"
defined "RangedSummarizedExperiment"
```

There is no `rownames()` method for SummarizedExperiment, which I agree is a bit confusing since SE has a matrix-like API. Are you suggesting this should be changed?

@lwaldron: The requirement for non-NULL `rownames()` seems quite strong to my way of working, but perhaps I've misunderstood. For example, when working with a RangedSummarizedExperiment I never explicitly set `names()` ("rownames") but rather think of the GRanges elements as being implicit IDs for each "row".  If forced to add rownames I'm likely to just use `as.character(seq_len(nrow(SE)))` or `as.character(rowRanges(SE))` and it seems to me I'm carrying these around the object for no material gain - I subset a RangedSummarizedExperiment based on the result of `findOverlaps()`-based operations and not these "rownames". Again, sorry if I've misunderstood the point under discussion.

**Comment by lwaldron** [comment]
We'll discuss in a few minutes in the code review, but in defining a minimum number of expectations for Elist elements, having them return a character vector of unique, valid R names in response to `rownames()` seems to me to be one of those requirements, for the following reasons:
- it is possible for all the supported Elist data classes to do this
- the returned vector can later be used to subset that same object unambiguously.  This seems necessary for MultiAssayView functionality, and probably other things

You should still be able to subset RSE using a GRanges object using findOverlaps, but it should also be possible to subset it using IDs returned by `rownames()`.

PS there is a rownames() method for SummarizedExperiment:

```
> example("SummarizedExperiment")
> rownames(se0) = 1:nrow(se0)
> rownames(se0)
  [1] "1"   "2"   "3"   "4"   "5"   "6"   "7"   "8"   "9"   "10"  "11"  "12"  "13"  "14"  "15"  "16"  "17" 
 [18] "18"  "19"  "20"  "21"  "22"  "23"  "24"  "25"  "26"  "27"  "28"  "29"  "30"  "31"  "32"  "33"  "34" 
(snip)
> class(se0)
[1] "SummarizedExperiment"
attr(,"package")
[1] "SummarizedExperiment"

**Comment by vjcitn** [comment]
On Wed, Feb 17, 2016 at 10:54 AM, Levi Waldron <notifications@github.com>
wrote:

> We'll discuss in a few minutes in the code review, but in defining a
> minimum number of expectations for Elist elements, having them return a
> character vector of unique, valid R names in response to rownames() seems
> to me to be one of those requirements, for the following reasons:
>
>    - it is possible for all the supported Elist data classes to do this
>
>
This would be a stretch for VcfStack, although if a named GRanges is bound
in to get a real RangedVcfStack then this would be true.  We want to
require very little of Elist elements up front or we will be inhibited from
allowing things like data base connections.

Revisiting the API document, we have

Other data classes can be used in the MultiAssayExperiment, as long as they
provide four methods: colnames(), rownames(), [i, j], and assay(), and
these methods behave as they do for RangedRaggedAssay(). See the Elist section
for more details.

At what point in the lifetime of an MAE do we need to have rownames()
yielding a vector of strings?  Not at construction, IMHO.  If there is an
Elist element that falls down when a character index is used with [, the
indexing fails and an error is thrown.  The user can then take steps to
ensure that the element has the necessary annotation before attempting the
indexing, but the infrastructure does not have to guarantee that this works
without annotation effort by the user.

**Comment by lwaldron** [comment]
We actually came to the same resolution by the end of the code review today, and the actual validity check is that those method exists, nothing about what they return. So I've updated the API, and your quoted sentence from the summary now ends at "Other data classes can be used in the `MultiAssayExperiment`, as long as they provide four methods: `colnames()`, `rownames()`, `[i, j]`, and `assay()`", and the [Elist validity](https://github.com/vjcitn/MultiAssayExperiment/wiki/MultiAssayExperiment-API#validity-1) section says:

* __Elist__
	* Ensure that method functions `[` (bracket), `colnames`, `rownames`, and `assay` are possible.

Any data class that provides the following methods can be used as an element of `Elist`. RangedSummarizedExperiment provides the template behavior for `Elist` elements, as follows. This is "template" behavior, but not explicit requirements:

- `colnames()`, by returning a character vector of sample identifiers
- `rownames()`, by returning a character vector of feature identifiers
- `[i, j]`, by returning the restriction of the instance to rows i and columns j
- `assay()`, by returning the numerical assay results in the form of a matrix

**Comment by lwaldron** [comment]
unlist/list method suggested by Martin has been implemented, check that the inefficiency is mitigated before closing this issue.

**Comment by LiNk-NY** [comment]
unlist/relist method implemented in 7d89eb0
MultiAssayExperiments created with real data see https://github.com/schifferl/biocMultiAssay/tree/master/vignettes (needs some updating)

**Comment by lwaldron** [comment]
I like this new unit test! For anyone looking only at this issue, it creates a shell class to RangedRaggedAssay, and defines only the methods that are supposed to be required, then makes sure it can be used in an Elist and MultiAssayExperiment:

```
<snip>
aShell <- new("MAEshell", a = rse)
newShell <- list(myShell = aShell)
<snip>
test_that("the methods check out", {
   expect_true(is(Elist(newShell), "Elist"))
   expect_true(is(MultiAssayExperiment(newShell, masPheno, rangemap), "MultiAssayExperiment"))
 })
```
